File: Tests/LedgelingsCoreTests/EdgeWorldTests.swift
import CoreGraphics
import Testing
@testable import LedgelingsCore

@Suite struct EdgeLoopTests {
    let loop = EdgeLoop(rect: CGRect(x: 10, y: 10, width: 1000, height: 600))

    @Test func runsCounterClockwiseFromTheBottomLeftCorner() {
        #expect(loop.length == 3200)
        #expect(loop.point(at: 500) == CGPoint(x: 510, y: 10))
        #expect(loop.point(at: 1300) == CGPoint(x: 1010, y: 310))
        #expect(loop.point(at: 2100) == CGPoint(x: 510, y: 610))
        #expect(loop.point(at: 2900) == CGPoint(x: 10, y: 310))
    }

    @Test func segmentsChangeExactlyAtCorners() {
        #expect(loop.segment(at: 999) == 0)
        #expect(loop.segment(at: 1000) == 1)
        #expect(loop.segment(at: 1600) == 2)
        #expect(loop.segment(at: 2600) == 3)
    }

    @Test func wrapsInBothDirections() {
        #expect(loop.wrap(3250) == 50)
        #expect(loop.wrap(-50) == 3150)
        #expect(loop.point(at: -300) == loop.point(at: 2900))
    }

    @Test func rotationStandsTheCreatureOnEachEdgeFacingInwards() {
        #expect(loop.rotation(ofSegment: 0) == 0)                       // floor
        #expect(abs(loop.rotation(ofSegment: 1) - .pi / 2) < 1e-9)      // right wall
        #expect(abs(loop.rotation(ofSegment: 2) - .pi) < 1e-9)          // ceiling
        #expect(abs(loop.rotation(ofSegment: 3) - 3 * .pi / 2) < 1e-9)  // left wall
        #expect(loop.inward(ofSegment: 0) == CGVector(dx: 0, dy: 1))
        #expect(loop.inward(ofSegment: 1) == CGVector(dx: -1, dy: 0))
    }

    @Test func findsTheNearestPointOnTheLoop() {
        let hit = loop.nearest(to: CGPoint(x: 400, y: 30))
        #expect(hit.t == 390)
        #expect(hit.distance == 20)
    }
}

@Suite struct EdgeWorldTests {
    @Test func oneScreenIsItsRectPulledInByTheInset() {
        let world = EdgeWorld(screens: [CGRect(x: 0, y: 0, width: 1020, height: 620)], inset: 10)
        #expect(world.loops == [EdgeLoop(rect: CGRect(x: 10, y: 10, width: 1000, height: 600))])
    }

    @Test func twoEqualScreensSideBySideBecomeOneLoopWithNoSeam() {
        let world = EdgeWorld(screens: [CGRect(x: 0, y: 0, width: 1000, height: 600),
                                        CGRect(x: 1000, y: 0, width: 1000, height: 600)], inset: 10)
        #expect(world.loops == [EdgeLoop(rect: CGRect(x: 10, y: 10, width: 1980, height: 580))])
    }

    /// A tall screen on the left, a short one on the right, floors level:
    ///
    ///     +------+
    ///     |      +------+
    ///     |      :      |
    ///     +------+------+
    @Test func aShorterNeighbourMakesAnOutsideCornerToWalkRound() {
        let world = EdgeWorld(screens: [CGRect(x: 0, y: 0, width: 1000, height: 800),
                                        CGRect(x: 1000, y: 0, width: 600, height: 500)], inset: 10)
        #expect(world.loops.count == 1)
        #expect(world.loops[0].vertices == [
            CGPoint(x: 10, y: 10), CGPoint(x: 1590, y: 10), CGPoint(x: 1590, y: 490),
            CGPoint(x: 990, y: 490),      // under the short screen's ceiling, reaching the tall one...
            CGPoint(x: 990, y: 790),      // ...then up the tall screen's wall
            CGPoint(x: 10, y: 790),
        ])
    }

    @Test func negativeCoordinatesWorkBecauseSecondaryScreensHaveThem() {
        let world = EdgeWorld(screens: [CGRect(x: 0, y: 0, width: 1000, height: 600),
                                        CGRect(x: -800, y: -200, width: 800, height: 600)], inset: 10)
        #expect(world.loops.count == 1)
        #expect(world.loops[0].vertices.count == 8)
        #expect(world.loops[0].vertices.first == CGPoint(x: -790, y: -190))
    }

    @Test func screensTouchingOnlyAtACornerStaySeparateLoops() {
        let world = EdgeWorld(screens: [CGRect(x: 0, y: 0, width: 1000, height: 600),
                                        CGRect(x: 1000, y: 600, width: 1000, height: 600)], inset: 10)
        #expect(world.loops.count == 2)
    }

    @Test func nearestPicksTheRightLoop() {
        let world = EdgeWorld(screens: [CGRect(x: 0, y: 0, width: 1000, height: 600),
                                        CGRect(x: 1000, y: 600, width: 1000, height: 600)], inset: 10)
        #expect(world.nearest(to: CGPoint(x: 1500, y: 1195)).loop == 1)
        #expect(world.nearest(to: CGPoint(x: 500, y: 0)).loop == 0)
    }

    @Test func anAbsurdInsetStillLeavesALoop() {
        let world = EdgeWorld(screens: [CGRect(x: 0, y: 0, width: 40, height: 40)], inset: 500)
        #expect(world.loops[0].length > 0)
    }
}

@Suite struct DayNightTests {
    let clock = DayNight(day: 180, night: 300)

    @Test func threeMinutesOfDayThenFiveOfNightForever() {
        #expect(!clock.isNight(at: 0))
        #expect(!clock.isNight(at: 179))
        #expect(clock.isNight(at: 180))
        #expect(clock.isNight(at: 479))
        #expect(!clock.isNight(at: 480))
        #expect(clock.isNight(at: 480 + 200))
    }

    @Test func skippingLandsAtTheStartOfTheOtherHalf() {
        #expect(clock.skippingToNextPhase(from: 50) == 180)
        #expect(clock.skippingToNextPhase(from: 200) == 480)
        #expect(clock.remaining(at: 200) == 280)
    }

    @Test func aZeroLengthNightMeansTheyNeverSleep() {
        #expect(!DayNight(day: 60, night: 0).isNight(at: 59.9))
        #expect(!DayNight(day: 60, night: 0).isNight(at: 60))
    }
}

File: win/LedgelingsCore.Tests/EdgeWorldTests.cs
namespace Ledgelings.Core.Tests;

public class EdgeLoopTests
{
    readonly EdgeLoop loop = new(new Rect(10, 10, 1000, 600));

    [Fact]
    public void RunsCounterClockwiseFromTheBottomLeftCorner()
    {
        Assert.Equal(3200, loop.Length);
        Assert.Equal(new Pt(510, 10), loop.Point(500));
        Assert.Equal(new Pt(1010, 310), loop.Point(1300));
        Assert.Equal(new Pt(510, 610), loop.Point(2100));
        Assert.Equal(new Pt(10, 310), loop.Point(2900));
    }

    [Fact]
    public void SegmentsChangeExactlyAtCorners()
    {
        Assert.Equal(0, loop.Segment(999));
        Assert.Equal(1, loop.Segment(1000));
        Assert.Equal(2, loop.Segment(1600));
        Assert.Equal(3, loop.Segment(2600));
    }

    [Fact]
    public void WrapsInBothDirections()
    {
        Assert.Equal(50, loop.Wrap(3250));
        Assert.Equal(3150, loop.Wrap(-50));
        Assert.Equal(loop.Point(2900), loop.Point(-300));
    }

    [Fact]
    public void RotationStandsTheCreatureOnEachEdgeFacingInwards()
    {
        Assert.Equal(0, loop.Rotation(0));                                  // floor
        Assert.True(Math.Abs(loop.Rotation(1) - Math.PI / 2) < 1e-9);       // right wall
        Assert.True(Math.Abs(loop.Rotation(2) - Math.PI) < 1e-9);           // ceiling
        Assert.True(Math.Abs(loop.Rotation(3) - 3 * Math.PI / 2) < 1e-9);   // left wall
        Assert.Equal(new Vec(0, 1), loop.Inward(0));
        Assert.Equal(new Vec(-1, 0), loop.Inward(1));
    }

    [Fact]
    public void FindsTheNearestPointOnTheLoop()
    {
        var hit = loop.Nearest(new Pt(400, 30));
        Assert.Equal(390, hit.T);
        Assert.Equal(20, hit.Distance);
    }
}

public class EdgeWorldTests
{
    [Fact]
    public void OneScreenIsItsRectPulledInByTheInset()
    {
        var world = new EdgeWorld(new[] { new Rect(0, 0, 1020, 620) }, 10);
        Assert.Equal(new[] { new EdgeLoop(new Rect(10, 10, 1000, 600)) }, world.Loops);
    }

    [Fact]
    public void TwoEqualScreensSideBySideBecomeOneLoopWithNoSeam()
    {
        var world = new EdgeWorld(new[] { new Rect(0, 0, 1000, 600), new Rect(1000, 0, 1000, 600) }, 10);
        Assert.Equal(new[] { new EdgeLoop(new Rect(10, 10, 1980, 580)) }, world.Loops);
    }

    /// <summary>A tall screen on the left, a short one on the right, floors level:
    /// <code>
    ///     +------+
    ///     |      +------+
    ///     |      :      |
    ///     +------+------+
    /// </code></summary>
    [Fact]
    public void AShorterNeighbourMakesAnOutsideCornerToWalkRound()
    {
        var world = new EdgeWorld(new[] { new Rect(0, 0, 1000, 800), new Rect(1000, 0, 600, 500) }, 10);
        Assert.Single(world.Loops);
        Assert.Equal(new[]
        {
            new Pt(10, 10), new Pt(1590, 10), new Pt(1590, 490),
            new Pt(990, 490),      // under the short screen's ceiling, reaching the tall one...
            new Pt(990, 790),      // ...then up the tall screen's wall
            new Pt(10, 790),
        }, world.Loops[0].Vertices);
    }

    [Fact]
    public void NegativeCoordinatesWorkBecauseSecondaryScreensHaveThem()
    {
        var world = new EdgeWorld(new[] { new Rect(0, 0, 1000, 600), new Rect(-800, -200, 800, 600) }, 10);
        Assert.Single(world.Loops);
        Assert.Equal(8, world.Loops[0].Vertices.Count);
        Assert.Equal(new Pt(-790, -190), world.Loops[0].Vertices[0]);
    }

    [Fact]
    public void ScreensTouchingOnlyAtACornerStaySeparateLoops()
    {
        var world = new EdgeWorld(new[] { new Rect(0, 0, 1000, 600), new Rect(1000, 600, 1000, 600) }, 10);
        Assert.Equal(2, world.Loops.Count);
    }

    [Fact]
    public void NearestPicksTheRightLoop()
    {
        var world = new EdgeWorld(new[] { new Rect(0, 0, 1000, 600), new Rect(1000, 600, 1000, 600) }, 10);
        Assert.Equal(1, world.Nearest(new Pt(1500, 1195)).Loop);
        Assert.Equal(0, world.Nearest(new Pt(500, 0)).Loop);
    }

    [Fact]
    public void AnAbsurdInsetStillLeavesALoop()
    {
        var world = new EdgeWorld(new[] { new Rect(0, 0, 40, 40) }, 500);
        Assert.True(world.Loops[0].Length > 0);
    }
}

public class DayNightTests
{
    readonly DayNight clock = new(180, 300);

    [Fact]
    public void ThreeMinutesOfDayThenFiveOfNightForever()
    {
        Assert.False(clock.IsNight(0));
        Assert.False(clock.IsNight(179));
        Assert.True(clock.IsNight(180));
        Assert.True(clock.IsNight(479));
        Assert.False(clock.IsNight(480));
        Assert.True(clock.IsNight(480 + 200));
    }

    [Fact]
    public void SkippingLandsAtTheStartOfTheOtherHalf()
    {
        Assert.Equal(180, clock.SkippingToNextPhase(50));
        Assert.Equal(480, clock.SkippingToNextPhase(200));
        Assert.Equal(280, clock.Remaining(200));
    }

    [Fact]
    public void AZeroLengthNightMeansTheyNeverSleep()
    {
        Assert.False(new DayNight(60, 0).IsNight(59.9));
        Assert.False(new DayNight(60, 0).IsNight(60));
    }
}
